src/HttpRequest.cc

```diff
@@ -739,3 +739,52 @@ HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntry
     }
 }
 
+/// a helper for validating FindListeningPortAddress()-found address candidates
+static const Ip::Address *
+FindListeningPortAddressInAddress(const Ip::Address *ip)
+{
+    // FindListeningPortAddress() callers do not want INADDR_ANY addresses
+    return (ip && !ip->isAnyAddr()) ? ip : nullptr;
+}
+
+/// a helper for handling PortCfg cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInPort(const AnyP::PortCfgPointer &port)
+{
+    return port ? FindListeningPortAddressInAddress(&port->s) : nullptr;
+}
+
+/// a helper for handling Connection cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInConn(const Comm::ConnectionPointer &conn)
+{
+    return conn ? FindListeningPortAddressInAddress(&conn->local) : nullptr;
+}
+
+const Ip::Address *
+FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry *ale)
+{
+    // Check all sources of usable listening port information, giving
+    // HttpRequest and masterXaction a preference over ALE.
+
+    const HttpRequest *request = callerRequest;
+    if (!request && ale)
+        request = ale->request;
+    if (!request)
+        return nullptr; // not enough information
+
+    const Ip::Address *ip = FindListeningPortAddressInPort(request->masterXaction->squidPort);
+    if (!ip && ale)
+        ip = FindListeningPortAddressInPort(ale->cache.port);
+
+    // XXX: also handle PROXY protocol here when we have a flag to identify such request
+    if (ip || request->flags.interceptTproxy || request->flags.intercepted)
+        return ip;
+
+    /* handle non-intercepted cases */
```

Comment:
This comment is rather misleading. This code path handles non-intercepted cases that lacked port info, not all non-intercepted cases. Ideally, all non-intercepted cases will be handled above this comment. I do not know how close we are to that ideal. Suggestions:

* handle non-intercepted cases that lacked explicit listening port info
* handle non-intercepted cases that were not handled above

Reply:
Well isn't it obvious by looking at the code that the later part handles whatever is remaining / not handled above it? I do not think we need to be so specific.

If a future developer can not read and understand that much code on his own then he should not be developing squid! :) (because that much expertise is expected)

Infact I was thinking of removing other comments too! (``// not enough information``, ``// may still be nil``). Because those are also obvious thing.

That function now has 9 lines of code and 6 lines of comments!! To me its an overkill.

But that said I will go ahead and modify the requested comment!

Reply:
> isn't it obvious by looking at the code that the later part handles whatever is remaining / not handled above it?

You have the wrong threat model :-). Allow me to explain:

For the sake of the argument, let's assume that the code above the comment is fairly obvious. Many developers *are not going to read that code* after seeing that comment. They are going to plug in their changes (targeting non-intercepted cases) under the comment, thanking the original authors for clearly identifying the right place for their changes! Next comes an overloaded reviewer. She looks at a `-U3` or similar diff which clearly identifies that the developer plugged their code in the right place. PR approved!

And yes, completely removing that misleading comment would be a viable solution as well.

I agree that the `not enough information` comment can be considered an overkill.

As for `may still be nil`, I find such comments helpful during bug triage. Yes, you can deduce that fact after analyzing a few code lines, but that takes a bit of time: The function code certainly tries very hard not to leave `ip` nil! Is it possible that all those attempts have failed? Let me analyze this helper function, which calls this helper function... Nil pointer dereference is an ever-present danger. When you have to look at many functions, hints like this do save time in my experience.

Said that, I agree that the `may still be nil` comment can be considered an overkill.